packages/block-editor/src/store/selectors.js

```diff
@@ -3071,6 +3071,10 @@ export const getBlockEditingMode = createRegistrySelector(
 				return 'disabled';
 			}
 
+			if ( state?.patternBlockEditingModes?.has( clientId ) ) {
+				return state?.patternBlockEditingModes.get( clientId );
+			}
```

Comment:
I see you mentioned in the description that you have follow-up for "zoom-out" editing modes too. I think we should ideally just have the selector return directly the state value in the end, but I'm ok with going in steps.

Reply:
One challenge is going to be that the selector depends on `editorMode` which is not within the reducer itself. So it's possible that in the end we'll have two values in the state for each block, a block editing mode for "write" and another one for "design" and the selector would pick the right value depending on the active mode.

Reply:
> What I find odd is that the denormalization only applies in the case of the "pattern block editing modes", why not just make a global state.blockEditingModes ?

Yeah, it's a tricky one. In the case of switching from design to write mode, we don't need to recalculate every pattern block, so storing it separately is useful in that case, we can retain the info across modes.

Similarly, we need to retain back compat for the block editing modes that might be set via `setBlockEditingMode`, so I think those also need to be stored in separate state so that we don't accidentally clear them when switching editor modes.

Reply:
oh yes, I mistakenly used the same name, we do need this to be separate from the original `setBlockEditingMode` state.

Reply:
> One challenge is going to be that the selector depends on editorMode which is not within the reducer itself.

My other PR does handle this by dispatching an action for switching editor mode to ensure the reducer re-runs:
https://github.com/WordPress/gutenberg/pull/66919/files#diff-d1fbd3af72ee0220c5c3ab2952a24d818174aed711b9a50b57c9544218a63988R1681

but it still reads the mode from the preferences store:
https://github.com/WordPress/gutenberg/pull/66919/files#diff-3c4b969f7547d02bc04a2c554e0f7d21da07c3f779714acf02800ee0341b50d9R279-R281

I'm not sure if this is the best approach. The reducer also needs access to the `blocks` store to check for `role: content` attributes, so it's not the only place where it selects from other stores and I saw there was some precedence for this in the block editor reducer.